### PR TITLE
Setup error screens

### LINKED ISSUE #1041: Hardware Setup Error Screens and Warning Messages
This will support messaging for:
- low-power + not plugged in.
- card reader not plugged in.

### PR DESCRIPTION
**Pull request by beausmith**
Fixes #1041 

This PR needs:
- [x] backend data to enable screens.
- [x] tests based upon mocked data.

# Error Screens

These screens will display for all users if the state is such.

## Low battery and no charger
image

## Card Reader not detected
image

## No Printer Detected
image

## Printer Error Detected
image

# Warnings on Insert Card Screen

These warnings are informational but non-blocking. They are visible to both Poll Workers and Voters.

Note: It is possible for both of these warnings to appear at the same time.

## No Power Warning
image

## No Accessible Controller Warning
image

### CODE REVIEW

File: src/AppRoot.tsx [hunk 1]
```diff
@@ -793,11 +841,30 @@ class AppRoot extends React.Component<Props, State> {
...
+      hasChargerAttached,
+      hasLowBattery,
+      hasPrinterAttached,
+      hasPrinterError,
       precinctId,
       tally,
       userSettings,
       votes,
     } = this.state
+    if (hasLowBattery && !hasChargerAttached) {
```
Comment (eventualbuddha) [inline src/AppRoot.tsx]: All these will stop a voter who is in the middle of voting, right? Is that the behavior we want?

File: src/AppRoot.tsx [hunk 2]
```diff
@@ -456,6 +471,37 @@ class AppRoot extends React.Component<Props, State> {
...
+          } catch (error) {
+            this.stopStatusPolling() // Assume backend is unavailable.
+          }
+        }
+      )
+    }
+  }
+
+  public stopStatusPolling = () => {
+    this.statusPoller && this.statusPoller.stop()
```
Comment (eventualbuddha) [inline src/AppRoot.tsx]:
```suggestion
    this.statusPoller?.stop()
```

File: src/AppRoot.tsx (hunk 1 above)
Comment (beausmith) [reply src/AppRoot.tsx]: Yes. For loss of any functionality which will not allow a voter to complete their vote, we display one of these error screens. 

In the case of this specific error, if the charger is not plugged in and `!hasLowBattery`, the voter will see a non-blocking notification warning (TBD) on the Insert Card screen which is informational is design and has copy like: "This machine is on battery power. Please ask a poll worker to plug-in the power cord for this machine."

File: src/AppRoot.tsx (hunk 2 above)
Comment (beausmith) [reply src/AppRoot.tsx]: Oh yeah! This was copy/pasta/edit, but should have caught it.

**Comment by beausmith** [comment]
@eventualbuddha - I have added a Hardware API. This seemed simpler than separate APIs for each. 

Open to all suggestions for any refactors, etc.

What else do you need from me to implement a "module-hardware" or whatever you think is best?

File: src/utils/Hardware.ts [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+  }
+
+  /**
+   * Reads Accessible Controller status
+   */
+  public async readAccesssibleControllerStatus(): Promise<
+    AcessibleControllerAPI
+  > {
+    console.log("Why is 'this' undefined?", this)
+    return this.accessibleControllerStatus
```
Comment (beausmith) [inline src/utils/Hardware.ts]: @eventualbuddha - The last commit on this branch is a WIP with debugging code. I'm stumped as to why `this` is undefined. `MemoryHardware` is modeled after the `MemoryCard`, but something must be different. Do you see where the error in this code is?

File: src/AppRoot.tsx [hunk 4]
```diff
@@ -456,6 +472,55 @@ class AppRoot extends React.Component<Props, State> {
...
+        async () => {
+          try {
+            // Possible implementation
+            const {
+              readAccesssibleControllerStatus,
+              readBatteryStatus,
+              readCardReaderStatus,
+              readPrinterStatus,
+            } = this.props.hardware
+            const accesssibleController = await readAccesssibleControllerStatus()
```
Comment (eventualbuddha) [inline src/AppRoot.tsx]: This is the call that’s making `this` be `undefined` inside `readAccesssibleControllerStatus`. When you call a function without a context in strict mode you get `this === undefined` (in non-strict mode you get `this === window`).

For example:
```ts
const obj = { m() { return this } }
const { m } = obj
const mBound = ::obj.m // https://github.com/tc39/proposal-bind-operator

// These all pass.
assert(obj.m() === obj)
assert(m() === undefined)
assert(m.call(obj) === obj)
assert(mBound() === obj)
```

Note that the bind operator proposal is still stage 0 and [won’t be implemented in TS until stage 3 at the earliest](https://github.com/Microsoft/TypeScript/issues/3508).

File: src/utils/Hardware.ts [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+  /**
+   * Sets Battery charging
+   */
+  setBatteryCharging(charging: boolean): Promise<void>
+
+  /**
+   * Sets Battery level.  Number between 0–1.
+   */
+  setBatteryLevel(level: number): Promise<void>
```
Comment (eventualbuddha) [inline src/utils/Hardware.ts]: These `set*` methods don’t represent operations we can actually perform in non-test scenarios, i.e. the battery level is what it is. I’d say these should not be present on the `Hardware` interface but should be defined on `MemoryHardware`. Unless I’m missing something?

File: src/utils/Hardware.ts (hunk 3 above)
Comment (eventualbuddha) [reply src/utils/Hardware.ts]: Commented at the source of the problem. It’s an unbound method problem. I’d fix it by doing `const { hardware } = this.props` and calling i.e. `await hardware.read…`.

**Comment by beausmith** [comment]
@eventualbuddha - I've rebased, squashed, and pushed commits. Everything is working, however there are a couple regressions I've failed to figure out. I've got to hang my hat for the night. Back to this tomorrow.

**Pull request merged**